File: csharp/CsSearch/CsSearch/SearchResult.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace CsSearch
{
	public class SearchResult
	{
		public Regex SearchPattern { get; private set; }
		public SearchFile File { get; set; }
		public int LineNum { get; private set; }
		public int MatchStartIndex { get; private set; }
		public int MatchEndIndex { get; private set; }
		public string Line { get; private set; }
		public IList<string> LinesBefore { get; private set; }
		public IList<string> LinesAfter { get; private set; }

		public SearchResult(Regex searchPattern, SearchFile file, int lineNum,
			int matchStartIndex, int matchEndIndex, string line)
		{
			Initialize(searchPattern, file, lineNum, matchStartIndex,
				matchEndIndex, line, new List<string>(), new List<string>());
		}

		public SearchResult(Regex searchPattern, SearchFile file, int lineNum, int matchStartIndex,
			int matchEndIndex, string line, IList<string> linesBefore, IList<string> linesAfter)
		{
			Initialize(searchPattern, file, lineNum, matchStartIndex,
				matchEndIndex, line, linesBefore, linesAfter);
		}

		private void Initialize(Regex searchPattern, SearchFile file, int lineNum, int matchStartIndex,
			int matchEndIndex, string line, IList<string> linesBefore, IList<string> linesAfter)
		{
			SearchPattern = searchPattern;
			File = file;
			LineNum = lineNum;
			MatchStartIndex = matchStartIndex;
			MatchEndIndex = matchEndIndex;
			Line = line;
			LinesBefore = linesBefore;
			LinesAfter = linesAfter;
		}

		public static int Compare(SearchResult r1, SearchResult r2)
		{
			if (r1 == null && r2 == null)
				return 0;
			if (r1 == null)
				return -1;
			if (r2 == null)
				return 1;
			if (r1.File != null && r2.File != null)
			{
				var sfCmp = SearchFile.Compare(r1.File, r2.File);
				if (sfCmp != 0)
				{
					return sfCmp;
				}
			}
			if (r1.LineNum == r2.LineNum)
			{
				return r1.MatchStartIndex - r2.MatchStartIndex;
			}
			return r1.LineNum - r2.LineNum;
		}

	}

	public class SearchResultsComparer : IComparer<SearchResult>
	{
		public int Compare(SearchResult r1, SearchResult r2)
		{
			return SearchResult.Compare(r1, r2);
		}
	}
}

File: kotlin/ktsearch/src/main/kotlin/ktsearch/SearchResult.kt
package ktsearch

/**
 * @author cary on 7/25/16.
 */
data class SearchResult(val searchPattern: Regex,
                        val file: SearchFile?,
                        val lineNum: Int,
                        val matchStartIndex: Int,
                        val matchEndIndex: Int,
                        val line: String,
                        val linesBefore: List<String>,
                        val linesAfter: List<String>) {

    private val noSearchFileText = "<text>"

    constructor(searchPattern: Regex,
                file: SearchFile?,
                lineNum: Int,
                matchStartIndex: Int,
                matchEndIndex: Int,
                line: String) :
            this(searchPattern, file, lineNum, matchStartIndex, matchEndIndex,
                    line, listOf(), listOf())

    override fun toString(): String {
        if (linesBefore.isNotEmpty() || linesAfter.isNotEmpty()) {
            return multiLineToString()
        } else {
            return singleLineToString()
        }
    }

    fun multiLineToString(): String {
        val lineSepLength = 80
        val fileString = if (file == null) noSearchFileText else file.toString()
        val sb = StringBuilder().
                append("=".repeat(lineSepLength)).append("\n").
                append(fileString).append(": ").append(lineNum).append(": [").
                append(matchStartIndex).append(":").append(matchEndIndex).
                append("]\n").append("-".repeat(lineSepLength)).append("\n")
        var currentLineNum = lineNum
        val lineNumPadding = (lineNum + linesAfter.size).toString().length
        val lineFormat = " %1$" + lineNumPadding + "d | %2\$s\n"
        if (linesBefore.isNotEmpty()) {
            currentLineNum -= linesBefore.size
            for (lineBefore in linesBefore) {
                sb.append(" ").append(String.format(lineFormat, currentLineNum,
                        lineBefore))
                currentLineNum++
            }
        }
        sb.append(">").append(String.format(lineFormat, lineNum, line))
        if (linesAfter.isNotEmpty()) {
            currentLineNum++
            for (lineAfter in linesAfter) {
                sb.append(" ").append(String.format(lineFormat, currentLineNum,
                        lineAfter))
                currentLineNum++
            }
        }
        return sb.toString()
    }

    private fun singleLineToString(): String {
        val sb = StringBuilder()
        if (file != null) {
            sb.append(file.toString())
        } else {
            sb.append(noSearchFileText)
        }

        if (lineNum == 0) {
            sb.append(" matches at [").
                    append(matchStartIndex).
                    append(":").
                    append(matchEndIndex).
                    append("]")
        } else {
            sb.append(": ").
                    append(lineNum).
                    append(": [").
                    append(matchStartIndex).
                    append(":").
                    append(matchEndIndex).
                    append("]: ").
                    append(formatMatchingLine(line))
        }
        return sb.toString()
    }

    private fun formatMatchingLine(line: String): String {
        return line.trim()
    }
}
